File: src/ERPTestingApp/Services/DataService.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using System.Text.Json;
using ERPTestingApp.Models.Masters;
using ERPTestingApp.Models.Enums;

namespace ERPTestingApp.Services
{
    /// <summary>
    /// In-memory data service with optional JSON persistence
    /// </summary>
    public class DataService : IDataService
    {
        private readonly ObservableCollection<ClassMaster> _classes = new();
        private readonly ObservableCollection<VAMaster> _vas = new();
        private readonly ObservableCollection<IECStandardMaster> _iecStandards = new();
        private readonly ObservableCollection<PartyMaster> _parties = new();
        private readonly ObservableCollection<LabelMaster> _labels = new();
        private readonly ObservableCollection<PartCodeMaster> _partCodes = new();
        private readonly ObservableCollection<ModelMaster> _models = new();

        private readonly string _dataPath;

        public DataService()
        {
            // Set data path in local app data
            var appDataPath = Path.Combine(
                Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                "ERPTestingApp"
            );
            Directory.CreateDirectory(appDataPath);
            _dataPath = Path.Combine(appDataPath, "data.json");

            // Load existing data or initialize seed data
            if (File.Exists(_dataPath))
            {
                LoadFromJson();
            }
            else
            {
                InitializeSeedData();
            }
        }

        #region Class Master
        public ObservableCollection<ClassMaster> GetAllClasses() => _classes;
        public ClassMaster? GetClassById(Guid id) => _classes.FirstOrDefault(c => c.Id == id);
        public void AddClass(ClassMaster classMaster) => _classes.Add(classMaster);
        public void UpdateClass(ClassMaster classMaster)
        {
            var existing = GetClassById(classMaster.Id);
            if (existing != null)
            {
                var index = _classes.IndexOf(existing);
                _classes[index] = classMaster;
            }
        }
        public void DeleteClass(Guid id)
        {
            var item = GetClassById(id);
            if (item != null) _classes.Remove(item);
        }
        #endregion

        #region VA Master
        public ObservableCollection<VAMaster> GetAllVAs() => _vas;
        public VAMaster? GetVAById(Guid id) => _vas.FirstOrDefault(v => v.Id == id);
        public void AddVA(VAMaster vaMaster) => _vas.Add(vaMaster);
        public void UpdateVA(VAMaster vaMaster)
        {
            var existing = GetVAById(vaMaster.Id);
            if (existing != null)
            {
                var index = _vas.IndexOf(existing);
                _vas[index] = vaMaster;
            }
        }
        public void DeleteVA(Guid id)
        {
            var item = GetVAById(id);
            if (item != null) _vas.Remove(item);
        }
        #endregion

        #region IEC Standard Master
        public ObservableCollection<IECStandardMaster> GetAllIECStandards() => _iecStandards;
        public IECStandardMaster? GetIECStandardById(Guid id) => _iecStandards.FirstOrDefault(i => i.Id == id);
        public void AddIECStandard(IECStandardMaster iecStandard) => _iecStandards.Add(iecStandard);
        public void UpdateIECStandard(IECStandardMaster iecStandard)
        {
            var existing = GetIECStandardById(iecStandard.Id);
            if (existing != null)
            {
                var index = _iecStandards.IndexOf(existing);
                _iecStandards[index] = iecStandard;
            }
        }
        public void DeleteIECStandard(Guid id)
        {
            var item = GetIECStandardById(id);
            if (item != null) _iecStandards.Remove(item);
        }
        #endregion

        #region Party Master
        public ObservableCollection<PartyMaster> GetAllParties() => _parties;
        public PartyMaster? GetPartyById(Guid id) => _parties.FirstOrDefault(p => p.Id == id);
        public void AddParty(PartyMaster party) => _parties.Add(party);
        public void UpdateParty(PartyMaster party)
        {
            var existing = GetPartyById(party.Id);
            if (existing != null)
            {
                var index = _parties.IndexOf(existing);
                _parties[index] = party;
            }
        }
        public void DeleteParty(Guid id)
        {
            var item = GetPartyById(id);
            if (item != null) _parties.Remove(item);
        }
        #endregion

        #region Label Master
        public ObservableCollection<LabelMaster> GetAllLabels() => _labels;
        public LabelMaster? GetLabelById(Guid id) => _labels.FirstOrDefault(l => l.Id == id);
        public void AddLabel(LabelMaster label) => _labels.Add(label);
        public void UpdateLabel(LabelMaster label)
        {
            var existing = GetLabelById(label.Id);
            if (existing != null)
            {
                var index = _labels.IndexOf(existing);
                _labels[index] = label;
            }
        }
        public void DeleteLabel(Guid id)
        {
            var item = GetLabelById(id);
            if (item != null) _labels.Remove(item);
        }
        #endregion

        #region Part Code Master
        public ObservableCollection<PartCodeMaster> GetAllPartCodes() => _partCodes;
        public PartCodeMaster? GetPartCodeById(Guid id) => _partCodes.FirstOrDefault(p => p.Id == id);
        public void AddPartCode(PartCodeMaster partCode)
        {
            // Populate navigation properties
            if (partCode.ClassId.HasValue)
            {
                var classItem = GetClassById(partCode.ClassId.Value);
                partCode.ClassName = classItem?.ClassName;
            }
            if (partCode.VAId.HasValue)
            {
                var vaItem = GetVAById(partCode.VAId.Value);
                partCode.VAName = vaItem?.VAName;
            }
            _partCodes.Add(partCode);
        }
        public void UpdatePartCode(PartCodeMaster partCode)
        {
            var existing = GetPartCodeById(partCode.Id);
            if (existing != null)
            {
                // Populate navigation properties
                if (partCode.ClassId.HasValue)
                {
                    var classItem = GetClassById(partCode.ClassId.Value);
                    partCode.ClassName = classItem?.ClassName;
                }
                if (partCode.VAId.HasValue)
                {
                    var vaItem = GetVAById(partCode.VAId.Value);
                    partCode.VAName = vaItem?.VAName;
                }
                var index = _partCodes.IndexOf(existing);
                _partCodes[index] = partCode;
            }
        }
        public void DeletePartCode(Guid id)
        {
            var item = GetPartCodeById(id);
            if (item != null) _partCodes.Remove(item);
        }
        #endregion

        #region Model Master
        public ObservableCollection<ModelMaster> GetAllModels() => _models;
        public ModelMaster? GetModelById(Guid id) => _models.FirstOrDefault(m => m.Id == id);
        public void AddModel(ModelMaster model)
        {
            // Populate navigation properties
            if (model.ClassId.HasValue)
            {
                var classItem = GetClassById(model.ClassId.Value);
                model.ClassName = classItem?.ClassName;
            }
            if (model.VAId.HasValue)
            {
                var vaItem = GetVAById(model.VAId.Value);
                model.VAName = vaItem?.VAName;
            }
            _models.Add(model);
        }
        public void UpdateModel(ModelMaster model)
        {
            var existing = GetModelById(model.Id);
            if (existing != null)
            {
                // Populate navigation properties
                if (model.ClassId.HasValue)
                {
                    var classItem = GetClassById(model.ClassId.Value);
                    model.ClassName = classItem?.ClassName;
                }
                if (model.VAId.HasValue)
                {
                    var vaItem = GetVAById(model.VAId.Value);
                    model.VAName = vaItem?.VAName;
                }
                var index = _models.IndexOf(existing);
                _models[index] = model;
            }
        }
        public void DeleteModel(Guid id)
        {
            var item = GetModelById(id);
            if (item != null) _models.Remove(item);
        }
        #endregion

        #region Seed Data
        private void InitializeSeedData()
        {
            // Class Master
            _classes.Add(new ClassMaster { ClassName = "Class 0.2" });
            _classes.Add(new ClassMaster { ClassName = "Class 0.5" });
            _classes.Add(new ClassMaster { ClassName = "Class 1.0" });
            _classes.Add(new ClassMaster { ClassName = "Class 3.0" });
            _classes.Add(new ClassMaster { ClassName = "Class 5.0" });

            // VA Master
            _vas.Add(new VAMaster { VAName = "5 VA" });
            _vas.Add(new VAMaster { VAName = "10 VA" });
            _vas.Add(new VAMaster { VAName = "15 VA" });
            _vas.Add(new VAMaster { VAName = "30 VA" });
            _vas.Add(new VAMaster { VAName = "50 VA" });

            // IEC Standard Master
            _iecStandards.Add(new IECStandardMaster { IECStandard = "IEC 60044-1" });
            _iecStandards.Add(new IECStandardMaster { IECStandard = "IEC 61869-2" });
            _iecStandards.Add(new IECStandardMaster { IECStandard = "IEC 60076" });

            // Party Master
            _parties.Add(new PartyMaster
            {
                PartyName = "ABC Electronics Ltd",
                ContactPerson = "John Smith",
                Email = "john.smith@abcelectronics.com",
                Address = "123 Industrial Area, Mumbai"
            });
            _parties.Add(new PartyMaster
            {
                PartyName = "XYZ Transformers",
                ContactPerson = "Sarah Johnson",
                Email = "sarah@xyztransformers.com",
                Address = "456 Power Street, Delhi"
            });
            _parties.Add(new PartyMaster
            {
                PartyName = "Global Power Systems",
                ContactPerson = "Michael Brown",
                Email = "michael.brown@globalpowersystems.com",
                Address = "789 Grid Road, Bangalore"
            });

            // Label Master
            _labels.Add(new LabelMaster { LabelType = LabelType.Square, LabelLink = "templates/square_label.pdf" });
            _labels.Add(new LabelMaster { LabelType = LabelType.Rect, LabelLink = "templates/rect_label.pdf" });

            // Part Code Master (with references)
            var class1 = _classes[0];
            var va1 = _vas[0];
            _partCodes.Add(new PartCodeMaster
            {
                PartCode = "PT-001",
                ItemDescription = "Potential Transformer 11kV",
                Ratio = 11000,
                ClassId = class1.Id,
                VAId = va1.Id,
                LabelType = LabelType.Square,
                ClassName = class1.ClassName,
                VAName = va1.VAName
            });

            // Model Master (with references)
            _models.Add(new ModelMaster
            {
                ModelName = "PT-11kV-S1",
                Ratio = 11000,
                ClassId = class1.Id,
                VAId = va1.Id,
                Phase = Phase.OnePhase,
                ClassName = class1.ClassName,
                VAName = va1.VAName
            });
        }
        #endregion

        #region Persistence
        public void SaveToJson()
        {
            try
            {
                var data = new
                {
                    Classes = _classes.ToList(),
                    VAs = _vas.ToList(),
                    IECStandards = _iecStandards.ToList(),
                    Parties = _parties.ToList(),
                    Labels = _labels.ToList(),
                    PartCodes = _partCodes.ToList(),
                    Models = _models.ToList()
                };

                var json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(_dataPath, json);
            }
            catch (Exception ex)
            {
                // Log error (in production, use proper logging)
                Console.WriteLine($"Error saving data: {ex.Message}");
            }
        }

        public void LoadFromJson()
        {
            try
            {
                var json = File.ReadAllText(_dataPath);
                var data = JsonSerializer.Deserialize<JsonElement>(json);

                // Load Classes
                if (data.TryGetProperty("Classes", out var classes))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<ClassMaster>>(classes.GetRawText()) ?? new())
                        _classes.Add(item);
                }

                // Load VAs
                if (data.TryGetProperty("VAs", out var vas))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<VAMaster>>(vas.GetRawText()) ?? new())
                        _vas.Add(item);
                }

                // Load IEC Standards
                if (data.TryGetProperty("IECStandards", out var iecStandards))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<IECStandardMaster>>(iecStandards.GetRawText()) ?? new())
                        _iecStandards.Add(item);
                }

                // Load Parties
                if (data.TryGetProperty("Parties", out var parties))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<PartyMaster>>(parties.GetRawText()) ?? new())
                        _parties.Add(item);
                }

                // Load Labels
                if (data.TryGetProperty("Labels", out var labels))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<LabelMaster>>(labels.GetRawText()) ?? new())
                        _labels.Add(item);
                }

                // Load Part Codes
                if (data.TryGetProperty("PartCodes", out var partCodes))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<PartCodeMaster>>(partCodes.GetRawText()) ?? new())
                        _partCodes.Add(item);
                }

                // Load Models
                if (data.TryGetProperty("Models", out var models))
                {
                    foreach (var item in JsonSerializer.Deserialize<List<ModelMaster>>(models.GetRawText()) ?? new())
                        _models.Add(item);
                }
            }
            catch (Exception ex)
            {
                // Log error and initialize seed data
                Console.WriteLine($"Error loading data: {ex.Message}");
                InitializeSeedData();
            }
        }
        #endregion
    }
}

File: web/src/services/DataService.ts
import type {
    AppData,
    ClassMaster,
    VAMaster,
    IECStandardMaster,
    PartyMaster,
    LabelMaster,
    PartCodeMaster,
    ModelMaster,
    TestingScreenData
} from '../types/models';
import {
    seedClasses,
    seedVAs,
    seedIECStandards,
    seedParties,
    seedLabels,
    seedPartCodes,
    seedModels,
    defaultTestingData
} from '../data/seedData';

const STORAGE_KEY = 'erp_testing_app_data';

// Generate unique ID
export const generateId = (): string => {
    return Date.now().toString(36) + Math.random().toString(36).substr(2);
};

// Initialize data with seed data or load from localStorage
export const initializeData = (): AppData => {
    const stored = localStorage.getItem(STORAGE_KEY);
    if (stored) {
        try {
            const parsed = JSON.parse(stored);
            // Convert date strings back to Date objects
            return {
                ...parsed,
                classes: parsed.classes.map((c: ClassMaster) => ({
                    ...c,
                    createdAt: new Date(c.createdAt),
                    updatedAt: new Date(c.updatedAt)
                })),
                vas: parsed.vas.map((v: VAMaster) => ({
                    ...v,
                    createdAt: new Date(v.createdAt),
                    updatedAt: new Date(v.updatedAt)
                })),
                iecStandards: parsed.iecStandards.map((i: IECStandardMaster) => ({
                    ...i,
                    createdAt: new Date(i.createdAt),
                    updatedAt: new Date(i.updatedAt)
                })),
                parties: parsed.parties.map((p: PartyMaster) => ({
                    ...p,
                    createdAt: new Date(p.createdAt),
                    updatedAt: new Date(p.updatedAt)
                })),
                labels: parsed.labels.map((l: LabelMaster) => ({
                    ...l,
                    createdAt: new Date(l.createdAt),
                    updatedAt: new Date(l.updatedAt)
                })),
                partCodes: parsed.partCodes.map((pc: PartCodeMaster) => ({
                    ...pc,
                    createdAt: new Date(pc.createdAt),
                    updatedAt: new Date(pc.updatedAt)
                })),
                models: parsed.models.map((m: ModelMaster) => ({
                    ...m,
                    createdAt: new Date(m.createdAt),
                    updatedAt: new Date(m.updatedAt)
                })),
                testingData: parsed.testingData ? {
                    ...parsed.testingData,
                    revisionDate: new Date(parsed.testingData.revisionDate),
                    createdAt: new Date(parsed.testingData.createdAt),
                    updatedAt: new Date(parsed.testingData.updatedAt)
                } : null
            };
        } catch {
            console.error('Failed to parse stored data, using seed data');
        }
    }

    // Return seed data
    return {
        classes: [...seedClasses],
        vas: [...seedVAs],
        iecStandards: [...seedIECStandards],
        parties: [...seedParties],
        labels: [...seedLabels],
        partCodes: [...seedPartCodes],
        models: [...seedModels],
        testingData: { ...defaultTestingData }
    };
};

// Save data to localStorage
export const saveData = (data: AppData): void => {
    localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
};

// CRUD operations for Classes
export const addClass = (data: AppData, item: Omit<ClassMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: ClassMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, classes: [...data.classes, newItem] };
    saveData(newData);
    return newData;
};

export const updateClass = (data: AppData, id: string, item: Partial<ClassMaster>): AppData => {
    const newData = {
        ...data,
        classes: data.classes.map(c => c.id === id ? { ...c, ...item, updatedAt: new Date() } : c)
    };
    saveData(newData);
    return newData;
};

export const deleteClass = (data: AppData, id: string): AppData => {
    const newData = { ...data, classes: data.classes.filter(c => c.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for VAs
export const addVA = (data: AppData, item: Omit<VAMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: VAMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, vas: [...data.vas, newItem] };
    saveData(newData);
    return newData;
};

export const updateVA = (data: AppData, id: string, item: Partial<VAMaster>): AppData => {
    const newData = {
        ...data,
        vas: data.vas.map(v => v.id === id ? { ...v, ...item, updatedAt: new Date() } : v)
    };
    saveData(newData);
    return newData;
};

export const deleteVA = (data: AppData, id: string): AppData => {
    const newData = { ...data, vas: data.vas.filter(v => v.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for IEC Standards
export const addIECStandard = (data: AppData, item: Omit<IECStandardMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: IECStandardMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, iecStandards: [...data.iecStandards, newItem] };
    saveData(newData);
    return newData;
};

export const updateIECStandard = (data: AppData, id: string, item: Partial<IECStandardMaster>): AppData => {
    const newData = {
        ...data,
        iecStandards: data.iecStandards.map(i => i.id === id ? { ...i, ...item, updatedAt: new Date() } : i)
    };
    saveData(newData);
    return newData;
};

export const deleteIECStandard = (data: AppData, id: string): AppData => {
    const newData = { ...data, iecStandards: data.iecStandards.filter(i => i.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for Parties
export const addParty = (data: AppData, item: Omit<PartyMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: PartyMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, parties: [...data.parties, newItem] };
    saveData(newData);
    return newData;
};

export const updateParty = (data: AppData, id: string, item: Partial<PartyMaster>): AppData => {
    const newData = {
        ...data,
        parties: data.parties.map(p => p.id === id ? { ...p, ...item, updatedAt: new Date() } : p)
    };
    saveData(newData);
    return newData;
};

export const deleteParty = (data: AppData, id: string): AppData => {
    const newData = { ...data, parties: data.parties.filter(p => p.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for Labels
export const addLabel = (data: AppData, item: Omit<LabelMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: LabelMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, labels: [...data.labels, newItem] };
    saveData(newData);
    return newData;
};

export const updateLabel = (data: AppData, id: string, item: Partial<LabelMaster>): AppData => {
    const newData = {
        ...data,
        labels: data.labels.map(l => l.id === id ? { ...l, ...item, updatedAt: new Date() } : l)
    };
    saveData(newData);
    return newData;
};

export const deleteLabel = (data: AppData, id: string): AppData => {
    const newData = { ...data, labels: data.labels.filter(l => l.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for Part Codes
export const addPartCode = (data: AppData, item: Omit<PartCodeMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: PartCodeMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, partCodes: [...data.partCodes, newItem] };
    saveData(newData);
    return newData;
};

export const updatePartCode = (data: AppData, id: string, item: Partial<PartCodeMaster>): AppData => {
    const newData = {
        ...data,
        partCodes: data.partCodes.map(pc => pc.id === id ? { ...pc, ...item, updatedAt: new Date() } : pc)
    };
    saveData(newData);
    return newData;
};

export const deletePartCode = (data: AppData, id: string): AppData => {
    const newData = { ...data, partCodes: data.partCodes.filter(pc => pc.id !== id) };
    saveData(newData);
    return newData;
};

// CRUD operations for Models
export const addModel = (data: AppData, item: Omit<ModelMaster, 'id' | 'createdAt' | 'updatedAt'>): AppData => {
    const newItem: ModelMaster = {
        ...item,
        id: generateId(),
        createdAt: new Date(),
        updatedAt: new Date()
    };
    const newData = { ...data, models: [...data.models, newItem] };
    saveData(newData);
    return newData;
};

export const updateModel = (data: AppData, id: string, item: Partial<ModelMaster>): AppData => {
    const newData = {
        ...data,
        models: data.models.map(m => m.id === id ? { ...m, ...item, updatedAt: new Date() } : m)
    };
    saveData(newData);
    return newData;
};

export const deleteModel = (data: AppData, id: string): AppData => {
    const newData = { ...data, models: data.models.filter(m => m.id !== id) };
    saveData(newData);
    return newData;
};

// Testing Screen Data operations
export const updateTestingData = (data: AppData, testingData: Partial<TestingScreenData>): AppData => {
    const newTestingData = data.testingData
        ? { ...data.testingData, ...testingData, updatedAt: new Date() }
        : { ...defaultTestingData, ...testingData, updatedAt: new Date() };
    const newData = { ...data, testingData: newTestingData };
    saveData(newData);
    return newData;
};

export const resetTestingData = (data: AppData): AppData => {
    const newData = { ...data, testingData: { ...defaultTestingData, id: generateId() } };
    saveData(newData);
    return newData;
};

// Email validation
export const isValidEmail = (email: string): boolean => {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
};
